### PR TITLE
Perf: Speed up header rendering by not needing to include level/scrip…

### PR DESCRIPTION
**Pull request by trevorb**
I was doing some poking around in newrelic and noticed loading the header view was ridiculously slow for some reason. 

image

https://rpm.newrelic.com/accounts/501463/applications/3787364/transactions?type=app#id=5b22436f6e74726f6c6c65722f7363726970745f6c6576656c732f73686f77222c22225d

I tracked the slowness down to a function in user_model that includes a bunch of models that we no longer use in the query. (I checked everywhere we call this method, which is only like 2 places). I removed the includes as well as the stage param which is also no longer used.  This change cuts the header view rendering time to practically nothing.

### CODE REVIEW

**Comment by philbogle** [comment]
If you run with "pp=enabled", you can look at the SQL queries run with and without this change and verify the reduction in queries.

I am wondering whether summarize_user_progress (used for the progress widget at the top of the page) benefits from the included level assocation (or perhaps the level association doesn't help because it only uses script_levels but not the levels; not sure if loading levels also caches the :through association.)

**Comment by trevorb** [comment]
Yep, I used pp=enabled to debug the performance and sql query issues.

The widget at the top of the page is whats rendered in the _header.html.haml file. The only place that summarize_user_progress used the level association was the line I changed:
```
-        user_data[:levels][sl.level.id] = {
+        user_data[:levels][sl.level_id] = {
```

So now it only uses the script level objects.  

On another note, we might need to do some more research into the performance of the sql -> ruby marshaling as the majority of the slowness was not actually from the sql query but bringing the data into the ruby representation.

**Comment by trevorb** [comment]
Before:
image
After:
image

**Comment by laurelfan** [comment]
Hm, would it be possible to optimize this further by getting the script from the cache?

**Comment by trevorb** [comment]
Yep, I removed the order clause and did the sorting in ruby which completely removes the query for script levels. 

 
screenshot 2015-09-18 14 23 03

**Comment by laurelfan** [comment]
The script_levels should already be sorted by position, is this not true?

For this particular usage, I'd be wary of what's going on with assigning sl.user_level, now that you are using the cache. Assigning things to model objects that aren't actually things that are saved to the database is kind of a weird thing to do anyway, is it possible to refactor to not do this?

**Comment by trevorb** [comment]
Ahh, that's an excellent point about assigning to the cached objects. I'll do some more refactoring.

File: dashboard/app/models/user.rb [hunk 1]
```diff
@@ -417,15 +417,11 @@ def user_level_for(script_level)
                         level_id: script_level.level_id)
   end
 
-  def levels_from_script(script, stage = nil)
+  def levels_from_script(script)
     ul_map = user_levels_by_level(script)
-    q = script.script_levels.includes(:level, :script, :stage).order(:position)
+    sls = script.script_levels.sort { |a, b| a.position <=> b.position }
```
Comment (laurelfan) [inline dashboard/app/models/user.rb]: if you do need to do this it's a bit more compact to say `script.script_levels.sort_by(&:position)` (if you want to get in to implementation details, sort_by is different from sort in that it makes a pass to transform the objects into the sort values, which in this case might be a minuscule performance hit, see the doc for Enumerable#sort_by)

**Comment by philbogle** [comment]
This looks good to me once Laurel's comments are addressed.

**Comment by trevorb** [comment]
FYI, this change went out yesterday afternoon.   You can see a pretty nice drop for header rendering from ~200ms to about ~50ms, which is still a bit slower than I was expecting.
image
https://rpm.newrelic.com/accounts/501463/applications/3787364/transactions#id=5b22436f6e74726f6c6c65722f7363726970745f6c6576656c732f73686f77222c22225d

**Pull request merged**